In [1]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc read-aloud"><h2>--- Day 1: Inverse Captcha ---</h2><p>The night before Christmas, one of Santa's Elves calls you in a panic. "The printer's broken! We can't print the <em>Naughty or Nice List</em>!" By the time you make it to <span title="Floor 17: cafeteria, printing department, and experimental organic digitization equipment.">sub-basement 17</span>, there are only a few minutes until midnight. "We have a big problem," she says; "there must be almost <em>fifty</em> bugs in this system, but nothing else can print The List. Stand in this square, quick! There's no time to explain; if you can convince them to pay you in <em class="star">stars</em>, you'll be able to--" She pulls a lever and the world goes blurry.</p>
<p>When your eyes can focus again, everything seems a lot more pixelated than before. She must have sent you inside the computer! You check the system clock: <em>25 milliseconds</em> until midnight. With that much time, you should be able to collect all <em class="star">fifty stars</em> by December 25th.</p>
<p>Collect stars by solving puzzles.  Two puzzles will be made available on each <s style="text-decoration-color:#fff;">day</s> millisecond in the Advent calendar; the second puzzle is unlocked when you complete the first.  Each puzzle grants <em class="star">one star</em>. Good luck!</p>
<p>You're standing in a room with "digitization quarantine" written in LEDs along one wall. The only door is locked, but it includes a small interface. "Restricted Area - Strictly No Digitized Users Allowed."</p>
<p>It goes on to explain that you may only leave by solving a <a href="https://en.wikipedia.org/wiki/CAPTCHA">captcha</a> to prove you're <em>not</em> a human. Apparently, you only get one millisecond to solve the captcha: too fast for a normal human, but it feels like hours to you.</p>
<p>The captcha requires you to review a sequence of digits (your puzzle input) and find the <em>sum</em> of all digits that match the <em>next</em> digit in the list. The list is circular, so the digit after the last digit is the <em>first</em> digit in the list.</p>
<p>For example:</p>
<ul>
<li><code>1122</code> produces a sum of <code>3</code> (<code>1</code> + <code>2</code>) because the first digit (<code>1</code>) matches the second digit and the third digit (<code>2</code>) matches the fourth digit.</li>
<li><code>1111</code> produces <code>4</code> because each digit (all <code>1</code>) matches the next.</li>
<li><code>1234</code> produces <code>0</code> because no digit matches the next.</li>
<li><code>91212129</code> produces <code>9</code> because the only digit that matches the next one is the last digit, <code>9</code>.</li>
</ul>
<p><em>What is the solution</em> to your captcha?</p>
</article>


In [8]:
tests = [
    {
        "name": "Example 1",
        "digits": "1122",
        "expected": 3,
    },
    {
        "name": "Example 2",
        "digits": "1111",
        "expected": 4,
    },
    {
        "name": "Example 3",
        "digits": "1234",
        "expected": 0,
    },
    {
        "name": "Example 4",
        "digits": "91212129",
        "expected": 9,
    },
]


def inverse_captcha(digits: str) -> int:
    ds = [int(d) for d in digits.strip()]
    n = len(ds)
    return sum(ds[i] for i in range(n) if ds[i] == ds[(i + 1) % n])


run_tests_params(inverse_captcha, tests)


Test Example 1 passed, for inverse_captcha.
Test Example 2 passed, for inverse_captcha.
Test Example 3 passed, for inverse_captcha.
Test Example 4 passed, for inverse_captcha.
Success


In [9]:
with open("../input/day1.txt") as f:
    print(f"Part I: {inverse_captcha(f.read())}")

Part I: 1029


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>1029</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>You notice a progress bar that jumps to 50% completion. Apparently, the door isn't yet satisfied, but it did emit a <em class="star">star</em> as encouragement. The instructions change:</p>
<p>Now, instead of considering the <em>next</em> digit, it wants you to consider the digit <em>halfway around</em> the circular list.  That is, if your list contains <code>10</code> items, only include a digit in your sum if the digit <code>10/2 = 5</code> steps forward matches it. Fortunately, your list has an even number of elements.</p>
<p>For example:</p>
<ul>
<li><code>1212</code> produces <code>6</code>: the list contains <code>4</code> items, and all four digits match the digit <code>2</code> items ahead.</li>
<li><code>1221</code> produces <code>0</code>, because every comparison is between a <code>1</code> and a <code>2</code>.</li>
<li><code>123425</code> produces <code>4</code>, because both <code>2</code>s match each other, but no other digit has a match.</li>
<li><code>123123</code> produces <code>12</code>.</li>
<li><code>12131415</code> produces <code>4</code>.</li>
</ul>
<p><em>What is the solution</em> to your new captcha?</p>
</article>

</main>


In [10]:
tests_partII = [
    {
        "name": "Example 1",
        "digits": "1212",
        "expected": 6,
    },
    {
        "name": "Example 2",
        "digits": "1221",
        "expected": 0,
    },
    {
        "name": "Example 3",
        "digits": "123425",
        "expected": 4,
    },
    {
        "name": "Example 4",
        "digits": "123123",
        "expected": 12,
    },
    {
        "name": "Example 5",
        "digits": "12131415",
        "expected": 4,
    },
]


def inverse_captcha_II(digits: str) -> int:
    ds = [int(d) for d in digits.strip()]
    n = len(ds)
    return sum(ds[i] for i in range(n) if ds[i] == ds[(i + n // 2) % n])


run_tests_params(inverse_captcha_II, tests_partII)


Test Example 1 passed, for inverse_captcha_II.
Test Example 2 passed, for inverse_captcha_II.
Test Example 3 passed, for inverse_captcha_II.
Test Example 4 passed, for inverse_captcha_II.
Test Example 5 passed, for inverse_captcha_II.
Success


In [11]:
with open("../input/day1.txt") as f:
    print(f"Part II: {inverse_captcha_II(f.read())}")

Part II: 1220


<link href="style.css" rel="stylesheet"></link>
<p>Your puzzle answer was <code>1220</code>.</p>
<p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>
